plugins/producer_plugin/test/test_read_only_trx.cpp

```diff
@@ -0,0 +1,147 @@
+   fc::temp_directory temp;
+   auto temp_dir_str = temp.path().string();
+   
+   std::promise<std::tuple<producer_plugin*, chain_plugin*>> plugin_promise;
+   std::future<std::tuple<producer_plugin*, chain_plugin*>> plugin_fut = plugin_promise.get_future();
+   std::thread app_thread( [&]() {
+      fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::debug);
+      std::vector<const char*> argv = {"test", "--data-dir", temp_dir_str.c_str(), "--config-dir", temp_dir_str.c_str()};
+      argv.insert( argv.end(), specific_args.begin(), specific_args.end() );
+      app->initialize<chain_plugin, producer_plugin>( argv.size(), (char**) &argv[0] );
+      app->startup();
+      plugin_promise.set_value( {app->find_plugin<producer_plugin>(), app->find_plugin<chain_plugin>()} );
+      app->exec();
+   } );
+
+   auto[prod_plug, chain_plug] = plugin_fut.get();
+   auto chain_id = chain_plug->get_chain_id();
+
+   std::deque<packed_transaction_ptr> trxs;
+   std::atomic<size_t> next_calls = 0;
+   std::atomic<size_t> num_posts = 0;
+   std::atomic<size_t> trace_with_except = 0;
+   std::atomic<bool> trx_match = true;
+   const size_t num_pushes = 2558;
+
+   for( size_t i = 1; i <= num_pushes; ++i ) {
+      auto ptrx = make_unique_trx( chain_id );
+      app->executor().post( priority::low, exec_queue::general, [ptrx, &next_calls, &num_posts, &trace_with_except, &trx_match, &trxs, &app, i]() {
+         ++num_posts;
+         bool return_failure_traces = true;
+         app->get_method<plugin_interface::incoming::methods::transaction_async>()(ptrx,
+            false, // api_trx
+            transaction_metadata::trx_type::read_only, // trx_type
+            return_failure_traces,
+            [ptrx, &next_calls, &trace_with_except, &trx_match, &trxs, return_failure_traces]
+            (const std::variant<fc::exception_ptr, transaction_trace_ptr>& result) {
+               if( !std::holds_alternative<fc::exception_ptr>( result ) && !std::get<chain::transaction_trace_ptr>( result )->except ) {
+                  if( std::get<chain::transaction_trace_ptr>( result )->id != ptrx->id() ) {
+                     elog( "trace not for trx ${id}: ${t}",
+                           ("id", ptrx->id())("t", fc::json::to_pretty_string(*std::get<chain::transaction_trace_ptr>(result))) );
+                     trx_match = false;
+                  }
+               } else if( !return_failure_traces && !std::holds_alternative<fc::exception_ptr>( result ) && std::get<chain::transaction_trace_ptr>( result )->except ) {
+                  elog( "trace with except ${e}",
+                        ("e", fc::json::to_pretty_string( *std::get<chain::transaction_trace_ptr>( result ) )) );
+                  ++trace_with_except;
+               }
+               ++next_calls;
+        });
+      });
+   }
+
+   // Wait long enough such that all transactions are executed
+   auto start = fc::time_point::now();
+   auto end = start + fc::milliseconds(1000);
```

Comment:
On my machine, for debug build, this was not enough time. Also looks like it failed in CI/CD on my branch because it ran out of time.

Instead you could check `next_calls` and reset your timer anytime it changes.

Reply:
Thanks. Initially I was waiting for `next_calls` to be `num_pushes`, but then thought that was not a real test. Will change.

Reply:
I changed to:
```
   auto start = fc::time_point::now();
   auto end = start + fc::seconds(10);
   while ( fc::time_point::now() < end ){
      if (next_calls == num_pushes) break;
      std::this_thread::sleep_for( 500000us);
   }
```
On my branch.

Reply:
Thanks. Just saw this. I committed a version similar to yours in spirit.😀